File: app/schemas/order.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class OrderBase(BaseModel):

    status: str = Field(
        default="PENDING",
        max_length=50,
    )

    payment_status: str = Field(
        default="UNPAID",
        max_length=50,
    )

class OrderCreate(OrderBase):

    customer_id: int

class OrderUpdate(BaseModel):

    status: str | None = None
    payment_status: str | None = None
    delivery_name: str | None = None
    delivery_phone: str | None = None
    delivery_address: str | None = None
    city: str | None = None
    state: str | None = None
    postal_code: str | None = None
    country: str | None = None
    tracking_number: str | None = None

class OrderPrivate(BaseModel):

    model_config = ConfigDict(
        from_attributes=True
    )

    id: int
    customer_id: int
    status: str
    payment_status: str
    total_amount: Decimal
    delivery_name: str | None
    delivery_phone: str | None
    delivery_address: str | None
    city: str | None
    state: str | None
    postal_code: str | None
    country: str
    tracking_number: str | None
    shipped_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/order.ts
import type { DeliveryStatus } from "./delivery";

export type OrderStatus =
  | "PENDING"
  | "PROCESSING"
  | "SHIPPED"
  | "COMPLETED"
  | "CANCELLED";

export type PaymentStatus = "UNPAID" | "PAID" | "FAILED" | "REFUNDED";

export interface Order {
  id: number;
  customer_id: number;
  status: OrderStatus;
  payment_status: PaymentStatus;
  subtotal: number | string;
  discount_amount: number | string;
  total_amount: number | string;
  delivery_name: string | null;
  delivery_phone: string | null;
  delivery_address: string | null;
  city: string | null;
  state: string | null;
  postal_code: string | null;
  country: string;
  tracking_number: string | null;
  shipped_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface OrderItem {
  id: number;
  order_id: number;
  product_id: number;
  quantity: number;
  unit_price: number | string;
  subtotal: number | string;
  discount_id: number | null;
  discount_name: string | null;
  discount_type: "PERCENTAGE" | "FIXED_AMOUNT" | "BUNDLE_PRICE" | null;
  discount_value: number | string | null;
  discount_bundle_quantity: number | null;
  discount_amount: number | string;
  total_amount: number | string;
}

export interface CreateOrderItemData {
  product_id: number;
  quantity: number;
}

export interface CreateOrderData {
  customer_id: number;
  items: CreateOrderItemData[];
  delivery_name?: string | null;
  delivery_phone?: string | null;
  delivery_address?: string | null;
  city?: string | null;
  state?: string | null;
  postal_code?: string | null;
  country?: string;
}

export interface UpdateOrderData {
  status?: OrderStatus;
  payment_status?: PaymentStatus;
  delivery_name?: string | null;
  delivery_phone?: string | null;
  delivery_address?: string | null;
  city?: string | null;
  state?: string | null;
  postal_code?: string | null;
  country?: string | null;
  tracking_number?: string | null;
}

export interface OrderQuoteItem {
  product_id: number;
  product_name: string;
  quantity: number;
  unit_price: number | string;
  subtotal: number | string;
  discount_id: number | null;
  discount_name: string | null;
  discount_type: "PERCENTAGE" | "FIXED_AMOUNT" | "BUNDLE_PRICE" | null;
  discount_value: number | string | null;
  discount_bundle_quantity: number | null;
  discount_amount: number | string;
  total_amount: number | string;
}

export interface OrderQuote {
  items: OrderQuoteItem[];
  subtotal: number | string;
  discount_amount: number | string;
  total_amount: number | string;
}

export type FulfillmentStage = "NEEDS_ATTENTION" | "READY_TO_PREPARE" | "IN_PREPARATION" | "IN_DELIVERY";

export interface FulfillmentDelivery {
  id: number;
  status: DeliveryStatus;
  courier: string | null;
  tracking_number: string | null;
  updated_at: string;
}

export interface FulfillmentOrder {
  id: number;
  customer_name: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  total_amount: number | string;
  created_at: string;
  queue_stage: FulfillmentStage;
  delivery: FulfillmentDelivery | null;
}

export interface FulfillmentQueue {
  items: FulfillmentOrder[];
  counts: Record<FulfillmentStage, number>;
}
